cmake/FindNVTX.cmake

```diff
@@ -0,0 +1,50 @@
+# (C) Copyright 2024- ECMWF.
+#
+# This software is licensed under the terms of the Apache Licence Version 2.0
+# which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
+# In applying this licence, ECMWF does not waive the privileges and immunities
+# granted to it by virtue of its status as an intergovernmental organisation
+# nor does it submit to any jurisdiction.
+
+if(CMAKE_C_COMPILER_ID STREQUAL "PGI" OR CMAKE_C_COMPILER_ID STREQUAL "NVHPC" )
+
+    set (DEFAULT_DR_HOOK_NVTX ON)
+
+    if( ${CMAKE_VERSION} VERSION_LESS "3.25" )
+        find_package(CUDAToolkit REQUIRED COMPONENTS CUDA::nvToolsExt)
+
+        find_path(NVTX_ROOT
+                NAMES include/nvToolsExt.h
+                HINTS ${CUDAToolkit_LIBRARY_DIR}/..
+                )
+
+        find_library(NVTX_LIBRARIES
+                NAMES libnvToolsExt.so nvToolsExt
+                HINTS ${NVTX_ROOT}/lib ${NVTX_ROOT}/lib64
+                )
+
+        find_path(NVTX_INCLUDE_DIRS
+                NAMES nvToolsExt.h
+                HINTS ${NVTX_ROOT}/include
+                )
+
+        include(FindPackageHandleStandardArgs)
+        find_package_handle_standard_args(NVTX DEFAULT_MSG
+                NVTX_LIBRARIES
+                NVTX_INCLUDE_DIRS
+                )
+
+        mark_as_advanced(
+                NVTX_LIBRARIES
+                NVTX_INCLUDE_DIRS
+        )
+    else()
+        find_package(CUDAToolkit REQUIRED COMPONENTS CUDA::nvtx3)
+        set (NVTX_LIBRARY "CUDA::nvtx3")
+        #TODO: Implementation for NVTX3
```

Comment:
Does this mean that with CMake >= 3.25 this feature is not yet implemented?

Reply:
That is correct. This is due to Nvidia messing with their libraries & creating [`nvtx3`](https://cmake.org/cmake/help/latest/module/FindCUDAToolkit.html#nvtx3) to replace the now deprecated [`nvToolsExt`](https://cmake.org/cmake/help/latest/module/FindCUDAToolkit.html#nvtoolsext). I'd originally decided against implementing nvtx3 functionality due to dependency hell, but checking again on our machine, we do actually have all the modules at the correct versions.

I take it you'd like me to take a look into this? All the framework is there, so _in theory_ it should be an easy job.

Reply:
If possible to use still use the deprecated  nvToolsExt with cmake >= 3.25 then this does not need to be done right now. I just would not expect a feature to stop working by just changing the CMake version.

Especially for GPU developments we like to use recent CMake versions that handle better OpenACC, CUDA, HIP; and it would be counterintuitive if this were to then exactly not work.

Reply:
I don't think it's possible to use the deprecated nvToolsExt with newer CMake versions. I'll see if I can get the alternative to work instead : )

Reply:
Done & passing tests on hpc2020 with nvtx3!